allow deref patterns to participate in exhaustiveness analysis

dianne opened the pull request:
Per [this proposal](https://hackmd.io/4qDDMcvyQ-GDB089IPcHGg#Exhaustiveness), this PR allows deref patterns to participate in exhaustiveness analysis. Currently all deref patterns enforce `DerefPure` bounds on their scrutinees, so this assumes all patterns it's analyzing are well-behaved. This also doesn't support [mixed exhaustiveness](https://hackmd.io/4qDDMcvyQ-GDB089IPcHGg#Mixed-exhaustiveness), and instead emits an error if deref patterns are used together with normal constructors. I think mixed exhaustiveness would be nice to have (especially if we eventually want to support arbitrary `Deref` impls[^1]), but it'd require more work to get reasonable diagnostics[^2].

Tracking issue for deref patterns: #87121

r? @Nadrieril 

[^1]: Regardless of whether we support limited exhaustiveness checking for untrusted `Deref` or always require other arms to be exhaustive, I think it'd be useful to allow mixed matching for user-defined smart pointers. And it'd be strange if it worked there but not for `Cow`.

[^2]: I think listing out witnesses of non-exhaustiveness can be confusing when they're not necessarily disjoint, and when you only need to cover some of them, so we'd probably want special formatting and/or explanatory subdiagnostics. And if it's implemented similarly to unions, we'd probably also want some way of merging witnesses; the way witnesses for unions can appear duplicated is pretty unfortunate. I'm not sure yet how the diagnostics should look, especially for deeply nested patterns.

dianne committed:
let deref patterns participate in usefulness/exhaustiveness

This does not yet handle the case of mixed deref patterns with normal
constructors; it'll ICE in `Constructor::is_covered_by`. That'll be
fixed in a later commit.

dianne committed:
add exhaustiveness/usefulness tests for deref patterns

dianne committed:
error early when mixing deref patterns with normal constructors

Without adding proper support for mixed exhaustiveness, mixing deref
patterns with normal constructors would either violate
`ConstructorSet::split`'s invariant 4 or 7. We'd either be ignoring rows
with normal constructors or we'd have problems in unspecialization from
non-disjoint constructors. Checking mixed exhaustivenss similarly to how
unions are currently checked should work, but the diagnostics for unions
are confusing. Since mixing deref patterns with normal constructors is
pretty niche (currently it only makes sense for `Cow`), emitting an
error lets us avoid committing to supporting mixed exhaustiveness
without a good answer for the diagnostics.

dianne commented on compiler/rustc_pattern_analysis/src/constructor.rs:
```diff
@@ -1039,15 +1049,27 @@ impl<Cx: PatCx> ConstructorSet<Cx> {
+        let mut deref_pat_present = false;
         for ctor in ctors.cloned() {
             match ctor {
+                DerefPattern(..) => {
+                    if !deref_pat_present {
+                        deref_pat_present = true;
+                        present.push(ctor);
+                    }
+                }
```
Am I right in assuming `rustc_pattern_analysis` is meant to compile without unstable features when the `rustc` feature isn't enabled? If not, this could be made a little prettier with `if_let_guards`.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14570346325/job/40866384622) [(plain)](https://github.com/rust-lang/rust/commit/ccd1c1f23d692d08a274707a08cfde893953f44c/checks/40866384622/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] dot test:false 0.103
[RUSTC-TIMING] ryu test:false 0.120
[RUSTC-TIMING] itoa test:false 0.077
   Compiling memchr v2.7.4
error[E0004]: non-exhaustive patterns: `&Constructor::DerefPattern(_)` not covered
   --> src/tools/rust-analyzer/crates/hir-ty/src/diagnostics/match_check/pat_analysis.rs:370:43
    |
370 |         let tys: SmallVec<[_; 2]> = match ctor {
    |                                           ^^^^ pattern `&Constructor::DerefPattern(_)` not covered
    |
note: `Constructor<MatchCheckCtx<'_>>` defined here
   --> /checkout/compiler/rustc_pattern_analysis/src/constructor.rs:677:1
    |
677 | pub enum Constructor<Cx: PatCx> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
702 |     DerefPattern(Cx::Ty),
    |     ------------ not covered
    = note: the matched value is of type `&Constructor<MatchCheckCtx<'_>>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
    |
423 ~             },
424 +             &Constructor::DerefPattern(_) => todo!()
    |

[RUSTC-TIMING] serde_repr test:false 0.620
    Checking lsp-types v0.95.0
[RUSTC-TIMING] memchr test:false 0.633
---
[RUSTC-TIMING] thiserror test:false 0.038
    Checking tracing-tree v0.3.1
[RUSTC-TIMING] tracing_tree test:false 0.146
    Checking windows v0.56.0
error[E0004]: non-exhaustive patterns: `&Constructor::DerefPattern(_)` not covered
   --> src/tools/rust-analyzer/crates/hir-ty/src/diagnostics/match_check/pat_analysis.rs:260:26
    |
260 |         let kind = match pat.ctor() {
    |                          ^^^^^^^^^^ pattern `&Constructor::DerefPattern(_)` not covered
    |
note: `Constructor<MatchCheckCtx<'_>>` defined here
   --> /checkout/compiler/rustc_pattern_analysis/src/constructor.rs:677:1
    |
677 | pub enum Constructor<Cx: PatCx> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
702 |     DerefPattern(Cx::Ty),
    |     ------------ not covered
    = note: the matched value is of type `&Constructor<MatchCheckCtx<'_>>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
    |
311 ~             },
312 +             &Constructor::DerefPattern(_) => todo!()
    |

error[E0004]: non-exhaustive patterns: `&Constructor::DerefPattern(_)` not covered
   --> src/tools/rust-analyzer/crates/hir-ty/src/diagnostics/match_check/pat_analysis.rs:334:15
    |
334 |         match ctor {
    |               ^^^^ pattern `&Constructor::DerefPattern(_)` not covered
    |
note: `Constructor<MatchCheckCtx<'_>>` defined here
   --> /checkout/compiler/rustc_pattern_analysis/src/constructor.rs:677:1
    |
677 | pub enum Constructor<Cx: PatCx> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
702 |     DerefPattern(Cx::Ty),
    |     ------------ not covered
    = note: the matched value is of type `&Constructor<MatchCheckCtx<'_>>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
    |
360 ~             },
361 +             &Constructor::DerefPattern(_) => todo!()
    |

For more information about this error, try `rustc --explain E0004`.
[RUSTC-TIMING] hir_ty test:false 5.318
error: could not compile `hir-ty` (lib) due to 3 previous errors
```

</details>

dianne commented on compiler/rustc_pattern_analysis/src/rustc.rs:
```diff
@@ -493,11 +494,14 @@ impl<'p, 'tcx: 'p> RustcPatCtxt<'p, 'tcx> {
-                fields = vec![];
-                arity = 0;
-                ctor = Opaque(OpaqueId::new());
+            PatKind::DerefPattern { subpattern, .. } => {
+                // NB(deref_patterns): This assumes the deref pattern is matching on a trusted
+                // `DerefPure` type. If the `Deref` impl isn't trusted, any deref pattern that can
+                // fail (possibly due to expanding or-patterns inside it) must not influence
+                // exhaustiveness analysis.
```
I'm assuming here that we're not trying to guarantee a single deref per column, but that'd also work if there's a practical way to do it. I'd have to look into how match lowering and or-pattern expansion work to see whether it's possible to at least guarantee it in that case. Of course, it's also consistent (and maybe easier to explain) if they can never participate in exhaustiveness for an untrusted `Deref` impl.

Nadrieril replied:
Not impossible to implement but not sure I'd want to guarantee that. Also this just cannot work with a hypothetical `DerefMove`: for that we necessarily need to call `deref` to determine the branch first. So I'm inclined to say "for deref patterns with custom Deref, we require `unsafe impl DerefPure`".

Nadrieril replied:
nit: "pattern that can fail" wasn't immediately clear to me. I might say "If the `Deref` impl isn't trusted, exhaustiveness must take into account that multiple calls to `deref` may return different results. Hence multiple `deref!` patterns cannot be exhaustive together unless each is exhaustive by itself."

dianne commented on compiler/rustc_pattern_analysis/src/rustc.rs:
```diff
@@ -1119,6 +1132,47 @@ pub fn analyze_match<'p, 'tcx>(
     Ok(report)
 }
 
+fn detect_mixed_deref_pat_ctors<'p, 'tcx>(
```
Is there a better home for this anywhere? The other use of `PatternColumn` I could find was in `lints.rs`, but this isn't really a lint, so it didn't feel right there.

dianne replied:
It's looking like maybe this should work for generic `PatCx`, so maybe it should live in `usefulness.rs`? I imagine r-a will need this too.

dianne replied:
I looked into rehousing this and making it work for generic `PatCx` on my local branch. It's a bit awkward, though: in order to avoid doing a pass for every single pattern, it'd still be the responsibility of the frontend to report when the pattern needs to be checked for mixed constructors (i.e. when a deref pattern was lowered), so the API is a bit awkward. Maybe an extra pass for every pattern is inconsequential perf though? I'm not sure what would be best, but I can open a perf experiment if needed. In the mean time, I'll think more about how to handle diagnostics for mixed exhaustiveness, to maybe avoid this check altogether.

Nadrieril replied:
Yeah, perf experiment seems the only way to know

Nadrieril replied:
As for where to house it, we could rename `lints.rs` to `checks.rs` and put it there?

dianne committed:
add stubbed-out cases for rust-analyzer

rust-analyzer doesn't construct `DerefPattern(_)` constructors, so these
shouldn't crash. It looks like this is how slice patterns are
implemented too.

bors commented:
:umbrella: The latest upstream changes (presumably #140415) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dianne commented:
Rebased to resolve the test/doctest conflicts.

Nadrieril commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 1a76c9dc1625bf45fd8d6e4797c5bdb90011a7a0 with merge 377ed5e3433b200e14f360ee68b384e0a3f066d8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1a76c9dc1625bf45fd8d6e4797c5bdb90011a7a0","merge_sha":"377ed5e3433b200e14f360ee68b384e0a3f066d8"} -->

Nadrieril commented on tests/ui/pattern/deref-patterns/usefulness/mixed-constructors.rs:
```diff
@@ -0,0 +1,48 @@
+        (0, Cow::Owned(_)) => {}
+        (0, Cow::Borrowed(_)) => {}
+        _ => {}
+        (0, true) => {}
```
To make sure someone doesn't move the check inside exhaustiveness checking:
```suggestion
        (1, true) => {}
```

Nadrieril reviewed (approved):
Pretty much flawless, can't find anything I'd change in this implementation. Started a perf run out of habit, not expecting a change. Happy to merge as-is, got some nits.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14869922522/job/41761131894)
Build commit: 377ed5e3433b200e14f360ee68b384e0a3f066d8 (`377ed5e3433b200e14f360ee68b384e0a3f066d8`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14869922522/job/41761131894"},"merge_sha":"377ed5e3433b200e14f360ee68b384e0a3f066d8"} -->

rust-timer commented:
Queued 377ed5e3433b200e14f360ee68b384e0a3f066d8 with parent 1a95cc6f9d950992ba27746e45caf40118581cb1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1a95cc6f9d950992ba27746e45caf40118581cb1&end=377ed5e3433b200e14f360ee68b384e0a3f066d8).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.8 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (377ed5e3433b200e14f360ee68b384e0a3f066d8): [comparison URL](https://perf.rust-lang.org/compare.html?start=1a95cc6f9d950992ba27746e45caf40118581cb1&end=377ed5e3433b200e14f360ee68b384e0a3f066d8&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=1a95cc6f9d950992ba27746e45caf40118581cb1&end=377ed5e3433b200e14f360ee68b384e0a3f066d8&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4% | [0.1%, 1.1%] | 4     |
| Regressions ❌ <br /> (secondary)  | 0.7% | [0.7%, 0.7%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.4% | [0.1%, 1.1%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1a95cc6f9d950992ba27746e45caf40118581cb1&end=377ed5e3433b200e14f360ee68b384e0a3f066d8&stat=max-rss)
<details>
<summary>Results (primary 0.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.4%, 0.7%]   | 6     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.7%, -0.6%] | 3     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.2%  | [-0.7%, 0.7%]  | 9     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1a95cc6f9d950992ba27746e45caf40118581cb1&end=377ed5e3433b200e14f360ee68b384e0a3f066d8&stat=cycles:u)
<details>
<summary>Results (primary -0.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.5%, 0.6%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.8%, -0.4%] | 9     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.4% | [-0.8%, 0.6%]  | 11    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=1a95cc6f9d950992ba27746e45caf40118581cb1&end=377ed5e3433b200e14f360ee68b384e0a3f066d8&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 769.964s -> 770.737s (0.10%)
**Artifact size**: 365.50 MiB -> 365.46 MiB (-0.01%)

dianne commented:
Applied the comment and test changes; ty! I also added a note about the mixed constructor check. I'll save moving it elsewhere and making it generic for later, since that's a bit more disruptive. [(diff)](https://github.com/rust-lang/rust/compare/1a76c9dc1625bf45fd8d6e4797c5bdb90011a7a0..2d4e19c5bb2e95ba754943ee9e34a617d4418c9b)

I'm not experienced with perf readouts, but that regression might be spurious? There's no regression in `check_match` afaict, and that should be all this affects. In the benchmark with the biggest change, `nalgebra`, it looks like the previous run was abnormally fast.

Nadrieril commented:
The perf diff is small enough to ignore imo. Let's merge!

@bors r+

bors commented:
:pushpin: Commit 09fed2d2f440fff2179ca9373bb16e40fc81d935 has been approved by `Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nadrieril 09fed2d2f440fff2179ca9373bb16e40fc81d935 -->
<!-- homu: {"type":"Approved","sha":"09fed2d2f440fff2179ca9373bb16e40fc81d935","approver":"Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 09fed2d2f440fff2179ca9373bb16e40fc81d935 with merge 7e552b46af72df390ed233b58a7f51650515b2a8...
<!-- homu: {"type":"BuildStarted","head_sha":"09fed2d2f440fff2179ca9373bb16e40fc81d935","merge_sha":"7e552b46af72df390ed233b58a7f51650515b2a8"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14897161036/job/41847940799)
Approved by: Nadrieril
Pushing 7e552b46af72df390ed233b58a7f51650515b2a8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Nadrieril","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14897161036/job/41847940799"},"merge_sha":"7e552b46af72df390ed233b58a7f51650515b2a8"} -->

rust-timer commented:
Finished benchmarking commit (7e552b46af72df390ed233b58a7f51650515b2a8): [comparison URL](https://perf.rust-lang.org/compare.html?start=ae3e8c6191fb2bf9394ea4201adaf7b1ac496120&end=7e552b46af72df390ed233b58a7f51650515b2a8&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ae3e8c6191fb2bf9394ea4201adaf7b1ac496120&end=7e552b46af72df390ed233b58a7f51650515b2a8&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ae3e8c6191fb2bf9394ea4201adaf7b1ac496120&end=7e552b46af72df390ed233b58a7f51650515b2a8&stat=max-rss)
<details>
<summary>Results (primary 0.6%, secondary -4.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4%  | [0.9%, 5.0%]   | 4     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-1.0%, -0.4%] | 6     |
| Improvements ✅ <br /> (secondary) | -4.2% | [-5.5%, -1.6%] | 4     |
| All ❌✅ (primary)                 | 0.6%  | [-1.0%, 5.0%]  | 10    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ae3e8c6191fb2bf9394ea4201adaf7b1ac496120&end=7e552b46af72df390ed233b58a7f51650515b2a8&stat=cycles:u)
<details>
<summary>Results (primary 0.7%, secondary 2.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7% | [0.4%, 1.3%] | 26    |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.2%, 2.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.7% | [0.4%, 1.3%] | 26    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ae3e8c6191fb2bf9394ea4201adaf7b1ac496120&end=7e552b46af72df390ed233b58a7f51650515b2a8&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 769.46s -> 769.201s (-0.03%)
**Artifact size**: 365.32 MiB -> 365.25 MiB (-0.02%)
